GH-36845: [C++][Python] Allow type promotion on `pa.concat_tables`

Fokko opened the pull request:
Revival of https://github.com/apache/arrow/pull/12000/

<!--
Thanks for opening a pull request!
If this is your first pull request you can find detailed information on how 
to contribute here:
  * [New Contributor's Guide](https://arrow.apache.org/docs/dev/developers/guide/step_by_step/pr_lifecycle.html#reviews-and-merge-of-the-pull-request)
  * [Contributing Overview](https://arrow.apache.org/docs/dev/developers/overview.html)


If this is not a [minor PR](https://github.com/apache/arrow/blob/main/CONTRIBUTING.md#Minor-Fixes). Could you open an issue for this pull request on GitHub? https://github.com/apache/arrow/issues/new/choose

Opening GitHub issues ahead of time contributes to the [Openness](http://theapacheway.com/open/#:~:text=Openness%20allows%20new%20users%20the,must%20happen%20in%20the%20open.) of the Apache Arrow project.

Then could you also rename the pull request title in the following format?

    GH-${GITHUB_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

or

    MINOR: [${COMPONENT}] ${SUMMARY}

In the case of PARQUET issues on JIRA the title also supports:

    PARQUET-${JIRA_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

-->

### Rationale for this change

It would be great to be able to do promotions when `concat`'ing a table, such as:

```python
def test_concat_tables_with_promotion_int():
    import pyarrow as pa
    t1 = pa.Table.from_arrays(
        [pa.array([1, 2], type=pa.int64())], ["int"])
    t2 = pa.Table.from_arrays(
        [pa.array([3, 4], type=pa.int32())], ["int"])

    result = pa.concat_tables([t1, t2], promote=True)

    assert result.equals(pa.Table.from_arrays([
        pa.array([1, 2, 3, 4], type=pa.int64())
    ], ["int"]))
```

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

### What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

### Are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

### Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please uncomment the line below and explain which changes are breaking.
-->
<!-- **This PR includes breaking changes to public APIs.** -->

<!--
Please uncomment the line below (and provide explanation) if the changes fix either (a) a security vulnerability, (b) a bug that caused incorrect or invalid data to be produced, or (c) a bug that causes a crash (even when the API contract is upheld). We use this to highlight fixes to issues that may affect users without their knowledge. For this reason, fixing bugs that cause errors don't count, since those are usually obvious.
-->
<!-- **This PR contains a "Critical Fix".** -->
* Closes: #36845

pitrou commented on cpp/src/arrow/type_test.cc:
```diff
@@ -983,6 +1091,242 @@ TEST_F(TestUnifySchemas, MoreSchemas) {
+  CheckUnify(decimal256(3, 2), {float32(), float64()}, options);
+
+  options.promote_integer_decimal = true;
+  CheckUnify(int32(), decimal128(3, 2), decimal128(3, 2), options);
```
Hmm, it's ok to risk truncation here? This is weird, as `int32` to `int16` isn't allowed.

Fokko replied:
This is not correct, and we should not allow truncation. Thanks for catching this!

pitrou commented on cpp/src/arrow/type_test.cc:
```diff
@@ -983,6 +1091,242 @@ TEST_F(TestUnifySchemas, MoreSchemas) {
+  CheckUnify(decimal128(3, 2), decimal128(5, 3), decimal128(5, 3), options);
+  CheckUnify(decimal128(3, 2), decimal128(5, 1), decimal128(6, 2), options);
+  CheckUnify(decimal128(3, 2), decimal128(5, -2), decimal128(9, 2), options);
+  CheckUnify(decimal128(3, -2), decimal128(5, -2), decimal128(5, -2), options);
```
What happens if I pass `decimal(38, 10)` and `decimal(38, 5)`. I expect it to fail. Can you add a test?

pitrou replied:
Also, should have similar tests for `decimal256`.

pitrou commented on cpp/src/arrow/type_test.cc:
```diff
@@ -983,6 +1091,242 @@ TEST_F(TestUnifySchemas, MoreSchemas) {
+  CheckUnify(int32(), decimal128(3, -2), decimal128(10, 0), options);
+
+  CheckUnifyFails(decimal256(1, 0), decimal128(1, 0), options);
+  CheckUnifyFails(int64(), decimal128(38, 37), options);
```
Add comments explaining why these would fail.

Fokko replied:
I think this one should pass right? `int64` is a `decimal(20, 0)`, which is compatible.

pitrou replied:
No, because the required precision (57 = 37+20) would be higher than what decimal128 allows.
However, perhaps it could upcast to decimal256 if the corresponding option is activated.

pitrou commented on cpp/src/arrow/type_test.cc:
```diff
@@ -983,6 +1091,242 @@ TEST_F(TestUnifySchemas, MoreSchemas) {
+             options);
+  CheckUnify(time32(TimeUnit::MILLI), {time64(TimeUnit::MICRO), time64(TimeUnit::NANO)},
+             options);
+  CheckUnify(time64(TimeUnit::MICRO), {time64(TimeUnit::NANO)}, options);
```
You should also check that unification fails when the relevant option isn't enabled.

pitrou commented on cpp/src/arrow/type_test.cc:
```diff
@@ -983,6 +1091,242 @@ TEST_F(TestUnifySchemas, MoreSchemas) {
+  CheckUnify(fixed_size_list(int8(), 2), {list(int8()), large_list(int8())}, options);
+
+  options.promote_nested = true;
+  CheckUnify(list(int8()), {list(int16()), list(int32()), list(int64())}, options);
```
Ok, but what if `promote_nested` is disabled?

Fokko replied:
I've removed promote nested

pitrou commented on cpp/src/arrow/type_test.cc:
```diff
@@ -983,6 +1091,242 @@ TEST_F(TestUnifySchemas, MoreSchemas) {
+  CheckUnify(list(int8()), {list(int16()), list(int32()), list(int64())}, options);
+  CheckUnify(fixed_size_list(int8(), 2),
+             {fixed_size_list(int16(), 2), list(int16()), list(int32()), list(int64())},
+             options);
```
What if I try to unify `fixed_size_list(int16, 2)` and `list(int8)`?

pitrou commented on cpp/src/arrow/type_test.cc:
```diff
@@ -983,6 +1091,242 @@ TEST_F(TestUnifySchemas, MoreSchemas) {
+
+TEST_F(TestUnifySchemas, Map) {
+  auto options = Field::MergeOptions::Defaults();
+  options.promote_nested = true;
```
Ok, but if `promote_nested` is false, what happens?

Fokko replied:
I've removed promote nested

pitrou commented on cpp/src/arrow/type_test.cc:
```diff
@@ -983,6 +1091,242 @@ TEST_F(TestUnifySchemas, MoreSchemas) {
+
+  ASSERT_RAISES(
+      Invalid,
+      field("foo", struct_({field("a", int8()), field("b", utf8()), field("a", int64())}))
+          ->MergeWith(field("foo", struct_({field("a", int64())})), options));
```
Since this is always tedious to write, I created https://github.com/apache/arrow/issues/36867

pitrou replied:
For the record, that PR was merged, so you can use it now.

pitrou commented on cpp/src/arrow/type_test.cc:
```diff
@@ -983,6 +1091,242 @@ TEST_F(TestUnifySchemas, MoreSchemas) {
+                 dictionary(int64(), utf8(), /*ordered=*/true),
+                 dictionary(int8(), large_utf8(), /*ordered=*/true),
+             },
+             options);
```
What if I try to unify `dictionary(int64(), utf8())` and `dictionary(int8(), large_utf8())`?

pitrou replied:
Ping on this.

pitrou commented on cpp/src/arrow/type_traits.h:
```diff
@@ -998,6 +998,21 @@ constexpr bool is_run_end_type(Type::type type_id) {
+///
+/// \param[in] type_id the type-id to check
+/// \return whether type-id is a primitive type one
+constexpr bool is_time(Type::type type_id) {
```
1) can you move this somewhere near `is_temporal`?
2) should we also define `is_date`?

Fokko replied:
1. Yes
2. Yes

pitrou commented on cpp/src/arrow/type_traits.h:
```diff
@@ -1187,7 +1202,6 @@ constexpr bool is_var_length_list(Type::type type_id) {
   switch (type_id) {
     case Type::LIST:
     case Type::LARGE_LIST:
-    case Type::MAP:
```
I don't think we want to change this?

pitrou replied:
Well, 1) It is a public API, so we probably shouldn't change the current semantics, 2) one should review internal uses of it to make sure changing it doesn't break anything that wouldn't be adequately tested.

pitrou commented on python/pyarrow/table.pxi:
```diff
@@ -5031,9 +5032,13 @@ def concat_tables(tables, c_bool promote=False, MemoryPool memory_pool=None):
+        See field_merge_options for the type promotion behavior.
     memory_pool : MemoryPool, default None
         For memory allocations, if required, otherwise use default pool.
+    field_merge_options : FieldMergeOptions, default None
```
Can we also allow passing "default" and "permissive" as strings?

Fokko replied:
I like it, added

pitrou commented on python/pyarrow/tests/test_schema.py:
```diff
@@ -725,6 +725,10 @@ def test_schema_merge():
     result = pa.unify_schemas((a, b, c))
     assert result.equals(expected)
 
+    result = pa.unify_schemas(
+        [b, d], options=pa.FieldMergeOptions.permissive())
+    assert result.equals(d)
```
It would be nice if I could also write:
```suggestion
    result = pa.unify_schemas(
        [b, d], options=pa.FieldMergeOptions.permissive())
    assert result.equals(d)
    result = pa.unify_schemas([b, d], options="permissive")
    assert result.equals(d)
```

@jorisvandenbossche What do you think?

Fokko replied:
I like it, added it!

pitrou commented on python/pyarrow/tests/test_table.py:
```diff
@@ -1373,6 +1373,15 @@ def test_concat_tables_with_promotion():
+        [pa.array([1, 2], type=pa.int32())], ["int64_field"])
+    result = pa.concat_tables(
+        [t1, t3], promote=True,
+        field_merge_options=pa.FieldMergeOptions.permissive())
```
Same here: would be nice to write `field_merge_options="permissive"`

pitrou commented on cpp/src/arrow/table_test.cc:
```diff
@@ -520,6 +522,36 @@ TEST_F(ConcatenateTablesWithPromotionTest, Simple) {
+  options.field_merge_options.promote_numeric_width = true;
+  ASSERT_OK_AND_ASSIGN(actual, ConcatenateTables({t1, t2}, options));
+  AssertTablesEqual(*expected_int64, *actual, /*same_chunk_layout=*/false);
+}
```
Can you tests for when casting fails because of overflow or other issues?

It would also showcase why the distinction between `TypeError` and `Invalid` is useful.

Fokko replied:
I wasn't sure what kind of tests you were looking for.

Fokko commented on cpp/src/arrow/type.cc:
```diff
@@ -316,6 +298,431 @@ std::shared_ptr<Field> Field::WithNullable(const bool nullable) const {
+        return uint16();
+      }
+      return uint8();
+    }
```
that's happening here:
```cpp
    if (left.id() == Type::DECIMAL256 || right.id() == Type::DECIMAL256 ||
        (options.promote_numeric_width &&
         common_precision > BasicDecimal128::kMaxPrecision)) {
      return DecimalType::Make(Type::DECIMAL256, common_precision, max_scale);
    }
```

Fokko commented on cpp/src/arrow/type.cc:
```diff
@@ -316,6 +298,431 @@ std::shared_ptr<Field> Field::WithNullable(const bool nullable) const {
+  if (maybe_promoted) return maybe_promoted;
+
+  if (options.promote_large) {
+    if (promoted_type->id() == Type::FIXED_SIZE_BINARY &&
```
I think we should remove `promote_large`, and replace it with `promote_binary`

Fokko commented on cpp/src/arrow/type.cc:
```diff
@@ -316,6 +298,431 @@ std::shared_ptr<Field> Field::WithNullable(const bool nullable) const {
+    if (promoted_type->id() == Type::FIXED_SIZE_BINARY &&
+        is_base_binary_like(other_type->id())) {
+      promoted_type = binary();
+      promoted = other_type->id() == Type::BINARY;
```
That does not make sense to me, I think it still should promote

Fokko commented on cpp/src/arrow/type.cc:
```diff
@@ -316,6 +298,431 @@ std::shared_ptr<Field> Field::WithNullable(const bool nullable) const {
+      promoted = promoted_type->id() == Type::BINARY;
+    }
+
+    if (promoted_type->id() == Type::FIXED_SIZE_LIST &&
```
This is for converting a fixed list into a variable-length list.

Fokko commented on cpp/src/arrow/type.cc:
```diff
@@ -316,6 +298,431 @@ std::shared_ptr<Field> Field::WithNullable(const bool nullable) const {
+        (promoted_type->id() == Type::LARGE_LIST && other_type->id() == Type::LIST)) {
+      promoted_type =
+          large_list(checked_cast<const BaseListType&>(*promoted_type).value_field());
+      promoted = true;
```
I took the original code from #12000. I would probably implement this using a post-order visitor pattern to traverse the schema. That would simplify the logic quite a bit.

Fokko commented on python/pyarrow/types.pxi:
```diff
@@ -1969,6 +1969,33 @@ cpdef KeyValueMetadata ensure_metadata(object meta, c_bool allow_none=False):
+        """
+        cdef FieldMergeOptions options = FieldMergeOptions()
+        options.c_options = CField.CMergeOptions.Permissive()
+        return options
```
That's a great suggestion

Fokko commented on cpp/src/arrow/type.cc:
```diff
@@ -327,17 +707,30 @@ Result<std::shared_ptr<Field>> Field::MergeWith(const Field& other,
-                         other.type()->ToString());
+    return std::make_shared<Field>(name_, promoted_type, nullable, metadata_);
+  }
+  return Status::TypeError("Unable to merge: Field ", name(),
```
I would argue that this would be a `TypeError`, but I'm afraid we can't just change this.

Fokko commented:
@github-actions autotune

Fokko commented:
@pitrou I addressed your comments. I appreciated your thorough review, would you be able to do another pass?

Fokko commented:
New week, new chances. Would you have time by any chance @pitrou ? :)

pitrou commented on cpp/src/arrow/type.cc:
```diff
@@ -316,6 +299,405 @@ std::shared_ptr<Field> Field::WithNullable(const bool nullable) const {
+    if (left.list_size() == right.list_size()) {
+      return fixed_size_list(std::move(value_field), left.list_size());
+    }
+    return Status::TypeError("Cannot merge fixed_size_list of different sizes");
```
Why not let them promote to a var-size list?
We promote if merging a fixed-size list and a var-size list, AFAICT.

Fokko replied:
Sounds reasonable, updated

pitrou commented on cpp/src/arrow/type.cc:
```diff
@@ -316,6 +299,405 @@ std::shared_ptr<Field> Field::WithNullable(const bool nullable) const {
+      return promoted_type;
+    }
+  } else if (promoted_type->id() == Type::NA || other_type->id() == Type::NA) {
+    return Status::Invalid("Cannot merge type with null unless promote_nullability=true");
```
`TypeError` here as well?

Fokko replied:
Yes, that's more appropriate

pitrou commented on cpp/src/arrow/type.cc:
```diff
@@ -327,14 +709,27 @@ Result<std::shared_ptr<Field>> Field::MergeWith(const Field& other,
+    } else if (nullable_ != other.nullable()) {
+      return Status::Invalid("Unable to merge: Field ", name(),
+                             " has incompatible nullability: ", nullable_, " vs ",
+                             other.nullable());
```
We should probably change these errors to `TypeError`. It may slightly break user code, but for the better (so that one can distinguish between unsupported type casts and invalid value casts).

pitrou commented on cpp/src/arrow/type_test.cc:
```diff
@@ -892,6 +892,150 @@ class TestUnifySchemas : public TestSchema {
+    CheckUnifyAsymmetric(field2, field1, expected, options);
+  }
+
+  void CheckUnifyFailsInvalid(
+      const std::shared_ptr<Field>& field1, const std::shared_ptr<Field>& field2,
+      const Field::MergeOptions& options = Field::MergeOptions::Defaults()) {
+    ARROW_SCOPED_TRACE("options: ", options);
+    ARROW_SCOPED_TRACE("field2: ", field2->ToString());
+    ARROW_SCOPED_TRACE("field1: ", field1->ToString());
+    ASSERT_RAISES(Invalid, field1->MergeWith(field2, options));
+    ASSERT_RAISES(Invalid, field2->MergeWith(field1, options));
+  }
```
I think it would be nicer to also check a part of the error message. For example (untested, but you get the idea):
```suggestion
  void CheckUnifyFailsInvalid(
      const std::shared_ptr<Field>& field1, const std::shared_ptr<Field>& field2,
      const Field::MergeOptions& options = Field::MergeOptions::Defaults(),
      const std::string& match_message = "") {
    ARROW_SCOPED_TRACE("options: ", options);
    ARROW_SCOPED_TRACE("field2: ", field2->ToString());
    ARROW_SCOPED_TRACE("field1: ", field1->ToString());
    EXPECT_RAISES_WITH_MESSAGE_THAT(
        Invalid, ::testing::HasSubstr(match_message), field1->MergeWith(field2, options));
    EXPECT_RAISES_WITH_MESSAGE_THAT(
        Invalid, ::testing::HasSubstr(match_message), field2->MergeWith(field1, options));
  }
```

pitrou commented on cpp/src/arrow/type_test.cc:
```diff
@@ -892,6 +892,150 @@ class TestUnifySchemas : public TestSchema {
+
+  void CheckUnifyFailsTypeError(
+      const std::shared_ptr<Field>& field1, const std::shared_ptr<Field>& field2,
+      const Field::MergeOptions& options = Field::MergeOptions::Defaults()) {
```
Same idea here :-)

pitrou commented on cpp/src/arrow/type_test.cc:
```diff
@@ -892,6 +892,150 @@ class TestUnifySchemas : public TestSchema {
+                         options);
+  }
+
+  void CheckUnify(const std::shared_ptr<DataType>& from,
```
For clarity, name it `CheckPromoteTo`? Otherwise it's a bit difficult to understand what happens when reading at the call site.

pitrou commented on cpp/src/arrow/type_test.cc:
```diff
@@ -892,6 +892,150 @@ class TestUnifySchemas : public TestSchema {
+    CheckUnifyFailsInvalid(field1, field2, options);
+  }
+
+  void CheckUnifyFailsInvalid(
```
Not suee this overload is useful since you can just pass both arguments as vectors?

pitrou commented on cpp/src/arrow/type_test.cc:
```diff
@@ -983,6 +1127,242 @@ TEST_F(TestUnifySchemas, MoreSchemas) {
+                           timestamp(TimeUnit::SECOND, "UTC"), options);
+
+  options.promote_temporal_unit = false;
+  CheckUnifyFailsInvalid(timestamp(TimeUnit::MICRO), timestamp(TimeUnit::NANO), options);
```
(would expect `TypeError` as well as in all/most instances here)

pitrou replied:
I mean that `Invalid` should be raised when a value fails casting, `TypeError` when we refuse casting from one type to another.

pitrou commented on cpp/src/arrow/type_test.cc:
```diff
@@ -983,6 +1127,242 @@ TEST_F(TestUnifySchemas, MoreSchemas) {
+  CheckUnifyFailsTypeError(fixed_size_list(int8(), 2), {fixed_size_list(int8(), 3)},
+                           options);
+
+  options.promote_binary = true;
```
I don't understand what `promote_binary` has to do with this?

Fokko replied:
No idea, removed it

pitrou commented on cpp/src/arrow/type.cc:
```diff
@@ -316,6 +299,405 @@ std::shared_ptr<Field> Field::WithNullable(const bool nullable) const {
+    const std::shared_ptr<DataType>& promoted_type,
+    const std::shared_ptr<DataType>& other_type, const Field::MergeOptions& options) {
+  bool promoted = false;
+
```
I would suggest a `promote_list` option to enable changing the list kind: for example from list to large_list. Doing it by default would be surprising to users IMHO.

The value type would always be promoted recursively, however.

Fokko replied:
I think if either of the sides is large, then I would expect large, but I can add this option

pitrou reviewed (changes requested):
Thanks for the update @Fokko . Here are some more comments. Also, it seems you need to rebase and fix conflicts?

Fokko commented on cpp/src/arrow/table.cc:
```diff
@@ -502,9 +504,16 @@ Result<std::shared_ptr<Table>> PromoteTableToSchema(const std::shared_ptr<Table>
-                           ": incompatible types: ", field->type()->ToString(), " vs ",
-                           current_field->type()->ToString());
+    if (!compute::CanCast(*current_field->type(), *field->type())) {
+      return Status::Invalid("Unable to promote field ", field->name(),
```
I went a bit back and forth here, but it looks like this one slipped through the cracks. I'm okay with changing the above test, but that will change the previous behavior. (but I think we're okay here)

Fokko commented on cpp/src/arrow/type.h:
```diff
@@ -397,14 +397,63 @@ class ARROW_EXPORT Field : public detail::Fingerprintable,
+    bool promote_numeric_width = false;
+
+    /// Allow strings to be promoted to binary types.
+    bool promote_binary = false;
```
I've updated the docstring

Fokko replied:
I've extended the docstring

Fokko commented on cpp/src/arrow/type.h:
```diff
@@ -397,14 +397,63 @@ class ARROW_EXPORT Field : public detail::Fingerprintable,
+    /// Second to millisecond, Time32 to Time64, Time32(SECOND) to Time32(MILLI), etc
+    bool promote_temporal_unit = false;
+
+    /// Promote dictionary index types to a common type, and unify the
+    /// value types.
```
Thank you, much better indeed

Fokko commented on cpp/src/arrow/type.h:
```diff
@@ -397,14 +397,63 @@ class ARROW_EXPORT Field : public detail::Fingerprintable,
+    /// value types.
+    bool promote_dictionary = false;
+
+    /// Allow merging ordered and non-ordered dictionaries, else
+    /// error. The result will be ordered if and only if both inputs
```
I agree. Explicit > implicit

Fokko commented on cpp/src/arrow/type.cc:
```diff
@@ -316,6 +299,405 @@ std::shared_ptr<Field> Field::WithNullable(const bool nullable) const {
+  ss << ", promote_dictionary=" << (promote_dictionary ? "true" : "false");
+  ss << ", promote_integer_sign=" << (promote_integer_sign ? "true" : "false");
+  ss << ", promote_binary=" << (promote_binary ? "true" : "false");
+  ss << '}';
```
Also the decimals we're missing, added those!

Fokko commented on python/pyarrow/types.pxi:
```diff
@@ -1969,6 +1969,48 @@ cpdef KeyValueMetadata ensure_metadata(object meta, c_bool allow_none=False):
+        return options
+
+    def __repr__(self) -> str:
+        return 'FieldMergeOptions(promote_nullability={},promote_decimal={},promote_decimal_to_float={},promote_integer_to_decimal={},promote_integer_to_float={},promote_integer_sign={},promote_numeric_width={},promote_binary={},promote_temporal_unit={},promote_dictionary={},promote_dictionary_ordered={})'.format(
```
Also, `promote_integer_to_float` was missing, added those as well 👍🏻

Fokko commented on cpp/src/arrow/type.cc:
```diff
@@ -316,6 +299,405 @@ std::shared_ptr<Field> Field::WithNullable(const bool nullable) const {
+      other_type->id() == Type::TIMESTAMP) {
+    const auto& left = checked_cast<const TimestampType&>(*promoted_type);
+    const auto& right = checked_cast<const TimestampType&>(*other_type);
+    if (left.timezone().empty() ^ right.timezone().empty()) {
+      return Status::TypeError(
+          "Cannot merge timestamp with timezone and timestamp without timezone");
+    }
+    if (left.timezone() != right.timezone()) {
+      return Status::TypeError("Cannot merge timestamps with differing timezones");
+    }
```
I think that's a good idea, updated the code 👍🏻

Fokko commented on cpp/src/arrow/type.cc:
```diff
@@ -316,6 +299,405 @@ std::shared_ptr<Field> Field::WithNullable(const bool nullable) const {
+                        right.precision() + max_scale - right.scale());
+  if (left.id() == Type::DECIMAL256 || right.id() == Type::DECIMAL256 ||
+      common_precision > BasicDecimal128::kMaxPrecision) {
+    return DecimalType::Make(Type::DECIMAL256, common_precision, max_scale);
```
This is already handled around line 515:
```c++
if (!options.promote_numeric_width && left.bit_width() != right.bit_width()) {
  return Status::TypeError(
      "Cannot promote decimal128 to decimal256 without promote_numeric_width=true");
}
```

Fokko replied:
Moved this to WidenDecimals.

Fokko commented on cpp/src/arrow/type.cc:
```diff
@@ -316,6 +299,405 @@ std::shared_ptr<Field> Field::WithNullable(const bool nullable) const {
+                                       (is_signed_integer(promoted_type->id()) &&
+                                        is_unsigned_integer(other_type->id())))) {
+    const int max_width =
+        std::max<int>(bit_width(promoted_type->id()), bit_width(other_type->id()));
```
Yes, I agree. I've added the following check:
```c++
if (bit_width(promoted_type->id()) != bit_width(other_type->id()) 
  && !options.promote_numeric_width) {
  return Status::TypeError(
      "Cannot widen an (un)signed integer without promote_numeric_width=true");
}
```

Fokko commented on cpp/src/arrow/type.cc:
```diff
@@ -316,6 +299,405 @@ std::shared_ptr<Field> Field::WithNullable(const bool nullable) const {
+                          MaxDecimalDigitsForInteger(other_type->id()));
+    ARROW_ASSIGN_OR_RAISE(const auto promoted_decimal,
+                          DecimalType::Make(promoted_type->id(), precision, 0));
+    ARROW_ASSIGN_OR_RAISE(promoted_type, WidenDecimals(promoted_type, promoted_decimal));
```
Yes, that's the best place to put it, thanks!

Fokko commented on cpp/src/arrow/type.cc:
```diff
@@ -316,6 +299,405 @@ std::shared_ptr<Field> Field::WithNullable(const bool nullable) const {
+      ((is_floating(promoted_type->id()) && is_integer(other_type->id())) ||
+       (is_integer(promoted_type->id()) && is_floating(other_type->id())))) {
+    const int max_width =
+        std::max<int>(bit_width(promoted_type->id()), bit_width(other_type->id()));
```
Great catch. I've updated the code.

Fokko commented on cpp/src/arrow/type.cc:
```diff
@@ -316,6 +299,405 @@ std::shared_ptr<Field> Field::WithNullable(const bool nullable) const {
+    } else if (promoted_type->id() == Type::FIXED_SIZE_BINARY &&
+               other_type->id() == Type::FIXED_SIZE_BINARY) {
+      return binary();
+    } else if (is_string(promoted_type->id()) && is_binary(other_type->id())) {
+      return MakeBinary(*other_type);
```
Should we merge those two together in the first place?

Fokko replied:
Removed them for now

Fokko replied:
Reinstated this, and added a swap to make sure that the large type is on the lefthand side.

Fokko commented on cpp/src/arrow/type.cc:
```diff
@@ -316,6 +299,405 @@ std::shared_ptr<Field> Field::WithNullable(const bool nullable) const {
+    } else {
+      return list(std::move(value_field));
+    }
+  } else if (promoted_type->id() == Type::STRUCT && other_type->id() == Type::STRUCT) {
```
I think splitting that out makes sense.

Fokko commented:
@pitrou Thanks for the reviews, I appreciate it. I wonder if we're not making this too big. Casting `int -> float`, `int -> decimal`, etc doesn't make much sense to me, and I don't think people expect much more than widening of the same type, such as `string -> large_string`, instead of also allowing `string -> binary`. I'm not interested in these special cases at least, it is just in here because the original author did this, but I'm also fine with only including the obvious and getting rid of all the flags. This also makes it much simpler, WDYT?

pitrou commented on cpp/src/arrow/type.cc:
```diff
@@ -316,6 +299,405 @@ std::shared_ptr<Field> Field::WithNullable(const bool nullable) const {
+  index_options.promote_integer_sign = true;
+  index_options.promote_numeric_width = true;
+  ARROW_ASSIGN_OR_RAISE(auto indices,
+                        MergeTypes(left.index_type(), right.index_type(), index_options));
```
Dictionaries in Arrow are like categoricals in Pandas. They are mostly for more efficient encoding of columns with many repetitions.

If you want flexible key-value mappings, a Map type should be used instead.

Fokko replied:
Got it, thanks for the context 👍🏻

Fokko commented:
@pitrou I took the existing code from https://github.com/apache/arrow/pull/12000 and went from there. For example, if feels to me that you always want to widen the types.

> Well, we shouldn't get rid of existing functionality. So any cast that is already supported should keep being supported (unless it's a glaring mistake of course!).

Well, many formats are much more restrictive than the first iteration here. Take for example Iceberg:

![image](https://github.com/apache/arrow/assets/1134248/c1451161-4996-4e6b-8966-539621cb7b75)

If you want to do anything special, such as combining a float and integer, you first need to cast it explicitly, which seems reasonable to me.

pitrou commented:
Ahah, I see. @lidavidm Could you take a look at @Fokko 's questions above?

lidavidm commented:
It's been long enough that I don't recall any particular reasoning, other than the cast being possible. You could certainly start smaller.

Though I don't believe dataset has any provision for manual casting, so if you do have (say) an int column that became a float column, this would be the only way to support it.

Fokko commented on cpp/src/arrow/table.h:
```diff
@@ -23,6 +23,7 @@
 #include <vector>
 
 #include "arrow/chunked_array.h"  // IWYU pragma: keep
+#include "arrow/compute/cast.h"
```
It does not seem to find the class when building the tests:
```
/usr/bin/ld: debug/libarrow.so.1300.0.0: undefined reference to `arrow::PromoteTableToSchema(std::shared_ptr<arrow::Table> const&, std::shared_ptr<arrow::Schema> const&, arrow::compute::CastOptions const&, arrow::MemoryPool*)'
collect2: error: ld returned 1 exit status
[319/835] Building CXX object src/arrow/compute/kernels/CMakeFiles/arrow-compute-scalar-if-else-test.dir/scalar_if_else_test.cc.o
ninja: build stopped: subcommand failed.
```
Any pointers?

Also, I think it would be nice to set the `CastOptions` to `::Safe()` by default, but I'm not able to get this working with the forward reference.

pitrou replied:
It's the _function_ `PromoteTableToSchema` that fails linking in your error message above. Meaning you probably didn't implement that particular signature?

> Also, I think it would be nice to set the `CastOptions` to `::Safe()` by default, but I'm not able to get this working with the forward reference.

Yes, this is why I suggested two signatures, one with `CastOptions` and one without.

Fokko replied:
Got it, thanks. I changed the `CastOptions` to a pointer because I can't forward declare the `CastOptions::Safe()` afaik. If there is a nullpointer, it will be set to `Safe()` within the `#ifdef ARROW_COMPUTE` block. Let me know what you think

jorisvandenbossche replied:
About including the compute headers here, @pitrou are you fine with the current state, or do you have another suggestion how to avoid it?

pitrou replied:
Still no fine, and the suggestion I made in https://github.com/apache/arrow/pull/36846/files/1c5643c4a855ca7a29fdbc4e3b83acd4e3d2af96#r1290117358 still holds.

Fokko replied:
Removed the import 👍  Digging into the failing CI

Fokko commented:
@lidavidm Thanks, in PyIceberg we do this by taking the field, casting them accordingly, and then turning it back into a table: https://github.com/apache/iceberg/blob/master/python/pyiceberg/io/pyarrow.py#L952-L1024 It is far from optimal, and the concat operation would be perfect. 

> Though I don't believe dataset has any provision for manual casting, so if you do have (say) an int column that became a float column, this would be the only way to support it.

The `pa.concat_tables` is about tables, and not about datasets. There is an option to cast them today: https://arrow.apache.org/docs/python/generated/pyarrow.Table.html#pyarrow.Table.cast

I think having the same safe flag on the `pa.concat_tables` would do the trick as well. WDYT @pitrou ?

lidavidm commented:
The underlying function in C++ is used in both contexts, and was my original motivation.

Fokko commented:
Alright, doing casts during reads has its [own issues](https://github.com/apache/arrow/issues/36845) (this might be faster, because it reads it right into the correct format?). Also, other folks are [waiting](https://github.com/apache/arrow/issues/37068) on this, so let's push this through. I now got my c++ environment running locally for faster development 👍🏻

Fokko commented on python/pyarrow/table.pxi:
```diff
@@ -5066,6 +5073,16 @@ def concat_tables(tables, c_bool promote=False, MemoryPool memory_pool=None):
+                options.field_merge_options = CField.CMergeOptions.Defaults()
+        else:
+            options.field_merge_options = cython.cast(
+                FieldMergeOptions, field_merge_options).c_options
```
Much better indeed, thanks!

Fokko commented on cpp/src/arrow/type.cc:
```diff
@@ -316,6 +299,405 @@ std::shared_ptr<Field> Field::WithNullable(const bool nullable) const {
+  } else if (values) {
+    return Status::Invalid("Could not merge index types");
+  }
+  return Status::Invalid("Could not merge value types");
```
Changed this to a `TypeError` as well

jorisvandenbossche commented:
@Fokko I needed this small diff to get pyarrow building locally:

```diff
diff --git a/python/pyarrow/types.pxi b/python/pyarrow/types.pxi
index 68efd1375..53b1fe65b 100644
--- a/python/pyarrow/types.pxi
+++ b/python/pyarrow/types.pxi
@@ -3239,13 +3239,12 @@ def unify_schemas(schemas, *, options=None):
         c_schemas.push_back(pyarrow_unwrap_schema(schema))
     if isinstance(options, str):
         if options == "default":
-            options = CField.CMergeOptions.Defaults()
+            c_options = CField.CMergeOptions.Defaults()
         elif options == "permissive":
-            options = CField.CMergeOptions.Permissive()
+            c_options = CField.CMergeOptions.Permissive()
         else:
             raise ValueError(f"Invalid merge option: {options}")
-
-    if options:
+    elif options is not None:
         c_options = (<FieldMergeOptions> options).c_options
     else:
         c_options = CField.CMergeOptions.Defaults()
```

Fokko commented:
@jorisvandenbossche Thanks for letting me know. I've pushed those changes. I'm working on the c++ tests to get them all happy and green

jorisvandenbossche commented on cpp/src/arrow/type.cc:
```diff
@@ -327,6 +315,449 @@ std::shared_ptr<Field> Field::WithNullable(const bool nullable) const {
+      // Other type is always the int
+      promoted_type.swap(other_type);
+    }
+    ARROW_ASSIGN_OR_RAISE(const int32_t precision,
+                          MaxDecimalDigitsForInteger(other_type->id()));
+    ARROW_ASSIGN_OR_RAISE(const auto promoted_decimal,
+                          DecimalType::Make(promoted_type->id(), precision, 0));
+    ARROW_ASSIGN_OR_RAISE(promoted_type,
+                          WidenDecimals(promoted_type, promoted_decimal, options));
```
~How useful is it to widen the decimal to the biggest possible precision needed to store the largest integer? Because often you have small integers. And you could also keep the existing decimal type, and rely on safe casting to catch errors?~

Actually, it seems our casting kernel has the same logic, so nevermind (it will error anyway)

jorisvandenbossche commented on cpp/src/arrow/type.cc:
```diff
@@ -327,6 +315,449 @@ std::shared_ptr<Field> Field::WithNullable(const bool nullable) const {
+    if (max_width >= 64) {
+      promoted_type = int64();
+    } else if (max_width >= 32) {
+      promoted_type = int32();
```
This means that `uint32` gets promoted to `int32`? But should that be `int64` instead?

jorisvandenbossche replied:
Did you address this comment? (`uint32` getting promoted to `int32`)

jorisvandenbossche commented on cpp/src/arrow/type.h:
```diff
@@ -397,14 +397,67 @@ class ARROW_EXPORT Field : public detail::Fingerprintable,
+    /// Second to millisecond, Time32 to Time64, Time32(SECOND) to Time32(MILLI), etc
+    bool promote_temporal_unit = false;
+
+    /// Allow promotion from a list to a large-list
+    bool promote_list = false;
```
And looking at the code, it also promotes fixed_sized_list of mixed length to variable sized list. That might be worth to mention in the doc comment above.

Fokko replied:
I agree, added!

jorisvandenbossche commented on python/pyarrow/types.pxi:
```diff
@@ -3173,6 +3215,9 @@ def unify_schemas(schemas):
     ----------
     schemas : list of Schema
         Schemas to merge into a single one.
+    options : FieldMergeOptions or string, optional
```
Maybe we should align the keyword name with the other places where we pass this? In `concat_tables`, it's called `field_merge_options`, which is quite long, so another option is something in between (`merge_options`, although that's probably not very descriptive being about the field types, or `promote_options`, but then that deviates from the naming in C++)

Fokko replied:
I can rename it to `promote_options` everywhere.

Fokko replied:
Hmm, maybe this 🚢 has sailed already. It is also called options in `cast(target_schema, safe, options)`, I like the consistency there

jorisvandenbossche replied:
Yes, but for the cast example, it are the CastOptions passed to the cast kernel. While here we have options for promoting types that is not tied to a specific function (i.e. it's used both in unify_schema, concat_tables, and potentially others)

jorisvandenbossche commented on python/pyarrow/types.pxi:
```diff
@@ -3173,6 +3215,9 @@ def unify_schemas(schemas):
     schemas : list of Schema
         Schemas to merge into a single one.
+    options : FieldMergeOptions or string, optional
+        Options for merging duplicate fields.
```
```suggestion
        Options for merging fields with the same name (which type to promote to).
```

("duplicate" is a bit confusing in this context I think)

jorisvandenbossche commented on python/pyarrow/types.pxi:
```diff
@@ -3186,12 +3231,26 @@ def unify_schemas(schemas):
+        else:
+            raise ValueError(f"Invalid merge option: {options}")
+
+    elif c_options:
```
```suggestion
    elif options:
```

jorisvandenbossche commented on python/pyarrow/types.pxi:
```diff
@@ -2014,6 +2014,48 @@ cpdef KeyValueMetadata ensure_metadata(object meta, c_bool allow_none=False):
+
+    __slots__ = ()
+
+    def __init__(self, *):
+        self.c_options = CField.CMergeOptions.Defaults()
```
Do we want to expose all fine grained options in Python? (now it's only the "permissive" and "default")

While trying this out, I expected I could do something like `concat_tables(..., field_merge_options=pa.FieldMergeOptions(promote_integer_to_decimal=True)`

If we only allow the two main options, I am not sure it is worth actually exposing this class publicly. 

(exposing the fine grained options in python can also be a follow-up)

jorisvandenbossche replied:
I am certainly fine with not exposing the fine-grained options for this PR. But in that case we maybe also don't need this class? (and just accept the string version in the function that have this option)

jorisvandenbossche commented on python/pyarrow/table.pxi:
```diff
@@ -5042,9 +5043,14 @@ def concat_tables(tables, c_bool promote=False, MemoryPool memory_pool=None):
     promote : bool, default False
-        If True, concatenate tables with null-filling and null type promotion.
+        If True, concatenate tables with null-filling and type promotion.
+        See field_merge_options for the type promotion behavior.
```
From a user experience perspective, I find it a bit strange that I still have to specify `promote=True` when I already specified field_merge_options="permissive" (because the latter only makes sense with promote=True). Essentially the `promote=False` can be seen as a "no promotion" field merge option? 

I know this is an existing keyword, but maybe going forward we should try to merge `promote` and `field_merge_options` in a single keyword? 

(on the naming side, I think I actually like "promote_options" better than "field_merge_options", but that would then deviate from the C++ naming ..)

Fokko replied:
> Essentially the promote=False can be seen as a "no promotion" field merge option?

This is correct, it will only do null-filling for missing columns.

The problem is that the `promote` is only relevant when merging tables, when just upcasting fields, then you don't need this, therefore `merge_options` might not be the best place to put it.

jorisvandenbossche replied:
> The problem is that the `promote` is only relevant when merging tables, when just upcasting fields, then you don't need this, therefore `merge_options` might not be the best place to put it.

Do you mean with "merging" tables for when the columns don't match? So if a column is not present in one of the tables, it gets filled with all nulls. This specific aspect is not something that you can control with `field_merge_options`? (you can only control whether to upcast null type to the target type, if the nulls where already present)

jorisvandenbossche reviewed (commented):
Great work @Fokko!

Added a bunch comments. But take them with a grain of salt: already a lot of work has gone in this PR, and not necessarily every comment necessarily needs to be addressed here (some things could be fine for follow-up issues as well)

Fokko commented on cpp/src/arrow/table.cc:
```diff
@@ -502,9 +515,23 @@ Result<std::shared_ptr<Table>> PromoteTableToSchema(const std::shared_ptr<Table>
+    columns.push_back(casted.chunked_array());
+#else
     return Status::Invalid("Unable to promote field ", field->name(),
                            ": incompatible types: ", field->type()->ToString(), " vs ",
```
Hmm, interesting. Looking at the PR, I think you're right!

Fokko commented on cpp/src/arrow/type.h:
```diff
@@ -397,14 +397,67 @@ class ARROW_EXPORT Field : public detail::Fingerprintable,
     /// nullable).
     bool promote_nullability = true;
 
+    /// Allow a decimal to be unified with another decimal of the same
+    /// width, adjusting scale and precision as appropriate. May fail
+    /// if the adjustment is not possible.
+    bool promote_decimal = false;
```
Good one, added that!

jorisvandenbossche replied:
I don't think one was added?

Fokko commented on cpp/src/arrow/table_test.cc:
```diff
@@ -520,6 +523,48 @@ TEST_F(ConcatenateTablesWithPromotionTest, Simple) {
+  AssertTablesEqual(*expected_int64, *actual, /*same_chunk_layout=*/false);
+}
+
+// TEST_F(ConcatenateTablesWithPromotionTest, Overflow) {
```
I've uncommented it. I think it is a good test to have

Fokko commented on cpp/src/arrow/type.h:
```diff
@@ -397,14 +397,67 @@ class ARROW_EXPORT Field : public detail::Fingerprintable,
+    /// accommodate the integer. (See increase_decimal_precision.)
+    bool promote_integer_to_decimal = false;
+
+    /// Allow an integer of a given bit width to be promoted to a
+    /// float; the result will be a float of an equal or greater bit
+    /// width to both of the inputs.
+    bool promote_integer_to_float = false;
```
Updated the docstring to:
```
/// Allow an integer of a given bit width to be promoted to a
/// float; the result will be a float of an equal or greater bit
/// width to both of the inputs. Examples:
///  - int8 + float32 = float32
///  - int32 + float32 = float64
///  - int32 + float64 = float64
/// Because an int32 cannot always be represented exactly in the 
/// 24 bits of a float32 mantissa.
```
Let me know if you had an other place in mind to document this.

Fokko commented on cpp/src/arrow/type.h:
```diff
@@ -397,14 +397,67 @@ class ARROW_EXPORT Field : public detail::Fingerprintable,
+    /// width to both of the inputs.
+    bool promote_integer_to_float = false;
+
+    /// Allow an unsigned integer of a given bit width to be promoted
+    /// to a signed integer of the equal or greater bit width.
+    bool promote_integer_sign = false;
```
This docstring is actually incorrect, it requires `promote_numeric_width` to be set to widen type type.

kou commented on c_glib/test/dataset/test-file-system-dataset-factory.rb:
```diff
@@ -109,9 +109,9 @@ def test_validate_fragments
+                "Type Error: Unable to merge: " +
                 "Field point has incompatible types: int16 vs int32"
-      error = assert_raise(Arrow::Error::Invalid) do
+      error = assert_raise(Arrow::Error::TypeError) do
```
```suggestion
      error = assert_raise(Arrow::Error::Type) do
```

Fokko replied:
Thank you @kou

kou commented on c_glib/test/dataset/test-file-system-dataset-factory.rb:
```diff
@@ -109,9 +109,9 @@ def test_validate_fragments
       options.validate_fragments = true
       message = "[file-system-dataset-factory][finish]: " +
-                "Invalid: Unable to merge: " +
+                "Type Error: Unable to merge: " +
```
```suggestion
                "Type error: Unable to merge: " +
```

Fokko replied:
Thank you @kou I'm not very much into the ruby ecosystem or its casing :)

kou replied:
Please ask me anything. :-)

Fokko commented:
Hey @pitrou and @jorisvandenbossche I was off last week, but it is all green again :) I went over all the comments (hope I didn't miss any). I saw the email today on the dev list on the feature freeze of 14.0.0. It would be awesome if we can get it in, do you see any chance?

Fokko commented:
It has been a week, gentle ping @pitrou and @jorisvandenbossche :)

jorisvandenbossche commented on cpp/src/arrow/table.h:
```diff
@@ -333,4 +342,29 @@ Result<std::shared_ptr<Table>> PromoteTableToSchema(
+/// the table, a column of nulls will be added to the resulting table.
+/// If the corresponding column is of type Null, it will be promoted
+/// to the type specified by schema, with null values filled. If Arrow
+/// was built with ARROW_COMPUTE, then the column will be casted to
```
same here

jorisvandenbossche commented on python/pyarrow/tests/test_table.py:
```diff
@@ -1330,6 +1330,22 @@ def test_concat_tables():
 
 
+def test_concat_tables_permissive():
+    t1 = pa.Table.from_arrays([list(range(10))], names=('a',))
+    t2 = pa.Table.from_arrays([list(range(10, 20))], names=('a',))
+
+    result = pa.concat_tables([t1, t2], field_merge_options="permissive")
```
I am not sure this is a perfect test for "permissive", because this specific case would perfectly work as well without specifying `field_merge_options="permissive"` (since the types are the same), and so it doesn't test that the option is properly passed through

Fokko replied:
Fair point, updated it to a rainy day situation

jorisvandenbossche commented on cpp/src/arrow/type.h:
```diff
@@ -397,14 +397,76 @@ class ARROW_EXPORT Field : public detail::Fingerprintable,
+    /// Allow an integer to be promoted to a decimal.
+    ///
+    /// May fail if the decimal has insufficient precision to
+    /// accommodate the integer. (See increase_decimal_precision.)
```
"increase_decimal_precision" doesn't exist? (should that be `promote_numeric_width`?)

jorisvandenbossche commented on cpp/src/arrow/type.h:
```diff
@@ -397,14 +397,76 @@ class ARROW_EXPORT Field : public detail::Fingerprintable,
+    /// to a signed integer that fits into the signed type:
+    /// uint + int16 = int16
+    /// When widening is needed, set promote_numeric_width to true:
+    /// uint16 + int16 = int32
```
This doesn't yet seem to be happening (related to an earlier comment), with the current branch:

```
In [37]: pa.unify_schemas([pa.schema([('a', 'uint16')]), pa.schema([('a', 'int16')])], options="permissive")
Out[37]: a: int16
```

Fokko replied:
That's correct, and has been fixed now 👍

jorisvandenbossche reviewed (commented):
@Fokko sorry for the delay! Slowly getting back into the PR

Something else for a follow-up: promoting bools to numeric?

We also have some logic around promoting types in kernels (`DispatchBest`). Longer term, we should probably also look into unifying that logic (the kernels can use the "permissive" promotion).

jorisvandenbossche commented:
Comparing the current branch with numpy promotion rules:

```python
import itertools
import pyarrow as pa
import numpy as np

types = ["uint8", "uint16", "uint32", "uint64", "int8", "int16", "int32", "int64", "float16", "float32", "float64"]

for t1, t2 in itertools.combinations(types, 2):
    np_result = np.result_type(t1, t2)
    pa_result = pa.unify_schemas([pa.schema([('a', t1)]), pa.schema([('a', t2)])], options="permissive").field('a').type
    pa_result_np = np.dtype(pa_result.to_pandas_dtype())
    if np_result != pa_result_np:
        print(f"{t1} + {t2}\t:\t{np_result} (numpy), {pa_result_np} (arrow)")
```
```
uint8 + int8	:	int16 (numpy), int8 (arrow)
uint16 + int8	:	int32 (numpy), int16 (arrow)
uint16 + int16	:	int32 (numpy), int16 (arrow)
uint32 + int8	:	int64 (numpy), int32 (arrow)
uint32 + int16	:	int64 (numpy), int32 (arrow)
uint32 + int32	:	int64 (numpy), int32 (arrow)
uint64 + int8	:	float64 (numpy), int64 (arrow)
uint64 + int16	:	float64 (numpy), int64 (arrow)
uint64 + int32	:	float64 (numpy), int64 (arrow)
uint64 + int64	:	float64 (numpy), int64 (arrow)
```

That shows the issue of widening the bitwidth for int+uint as commented above. 
Numpy also seems to go to float when combining uint64 with any signed integer (since uint64 values cannot all be held in int64). Not sure if we should also do that ..

And comparing with our numeric kernels DispatchBest (implemented in `CommonNumeric`, added in https://github.com/apache/arrow/pull/9294):

```python
for t1, t2 in itertools.combinations(types, 2):
    result_merge = pa.unify_schemas([pa.schema([('a', t1)]), pa.schema([('a', t2)])], options="permissive").field('a').type
    try:
        result_dispatch= pa.compute.add(pa.array([1, 2, 3], t1), pa.array(np.array([1, 2, 3], t2), t2)).type
    except:
        continue
    if result_merge != result_dispatch:
        print(f"{t1} + {t2}\t:\t{result_merge} (MergeOptions), {result_dispatch} (DispatchBest)")
```
```
uint8 + int8	:	int8 (MergeOptions), int16 (DispatchBest)
uint16 + int8	:	int16 (MergeOptions), int32 (DispatchBest)
uint16 + int16	:	int16 (MergeOptions), int32 (DispatchBest)
uint32 + int8	:	int32 (MergeOptions), int64 (DispatchBest)
uint32 + int16	:	int32 (MergeOptions), int64 (DispatchBest)
uint32 + int32	:	int32 (MergeOptions), int64 (DispatchBest)
uint32 + float32	:	double (MergeOptions), float (DispatchBest)
uint64 + float32	:	double (MergeOptions), float (DispatchBest)
int32 + float32	:	double (MergeOptions), float (DispatchBest)
int64 + float32	:	double (MergeOptions), float (DispatchBest)
```

So the numeric kernels already widen the uint+int case. And they do not widen the float bitwidth in the uint+float case (that's maybe something to change in DispatchBest in a follow-up).

pitrou commented:
Widening int and float types on arithmetic is really a mistake from NumPy, IMHO, so I don't think we should imitate it.

Fokko commented:
I'm not sure what you're referring to here: https://github.com/apache/arrow/pull/36846#discussion_r1336872769

This one is being used here:

```cpp
if (options.promote_decimal && is_decimal(promoted_type->id()) &&
  is_decimal(other_type->id())) {
ARROW_ASSIGN_OR_RAISE(promoted_type,
                      WidenDecimals(promoted_type, other_type, options));
return promoted_type;
}
```

Fokko commented:
I already implemented the logic of combining uints and ints, but there was a bug. I've fixed it and added some tests as well.

> Numpy also seems to go to float when combining uint64 with any signed integer (since uint64 values cannot all be held in int64). Not sure if we should also do that ..

I agree, that's a bit awkward. If it doesn't fit, we'll get an overflow exception.

jorisvandenbossche commented:
About https://github.com/apache/arrow/pull/36846#issuecomment-1735513388:

> I'm not sure what you're referring to here: [#36846 (comment)](https://github.com/apache/arrow/pull/36846#discussion_r1336872769)

I suggested to add that to the doc comment of `promote_decimal` (like you did in the docstring for `promote_integer_sign`):

> Ah, I see in the code that this can be done with `promote_numeric_width`. Maybe mention that here in the doc comment as well? (I didn't directly think about that)

jorisvandenbossche commented:
@Fokko thanks for the updates!

Going throught my previous reviews, summarizing here:

Some follow-up ideas that I don't think need to be addressed here:
- Mismatching timezones: a potential option could be to actually allow promoting timezones, and with the result using UTC (the actual time points will stay the same since we store UTC under the hood, you only loose the information about the specified timezone)
- Promoting bools to numeric?
- Exposing the fine-grained options in python (if we actually want to do this?)
- We have some logic around promoting types in kernels (`DispatchBest`, implemented in `CommonNumeric` for numeric types, added in #9294). We should look into consolidating that logic (the kernels can use the "permissive" promotion).


Then there are a few things that I would prefer to resolve here in this PR:

- Do we need the `pyarrow.FieldMergeOptions` class? At the moment it actually doesn't really do anything, since we don't (yet) allow to tweak individual flags, only "permissive"/"default" (and the class itself actually also only exposes "permissive", not "default", as a static method to create that options object). Maybe we can just remove it for now, we can always add it back later if we want to expose individual options.  
  I just noticed that the repr is also failing, so if we keep it, we should fix and test that. 


- For `concat_tables`, I find it confusing that a user still has to specify `promote=True` when they already specified `field_merge_options="permissive"` (because the latter only makes sense with promote=True). Essentially the `promote=False` can be seen as a "no promotion" field merge option?  
  I think with this new promotion capability, the historically added `promote` keyword is unfortunately badly chosen. Essentially it means: "unify schemas" (that's also how it is called on the C++ side), and by default it hardly does any promotion (only non-existing fields or null fields to any type). If we keep `promote` keyword, I would expect that `promote=True` would do actual, permissive promotion. 
  
  To be clear, this is not caused by this PR, but I would still prefer to get a nicer API. One option is to actually rename promote to `unify_schemas` (which is what it does). We can still keep `promote` as an alias for now for backwards compatibility. 
  
  And personally, I am also wondering if we shouldn't make "permissive" the default, when schema unification is enabled (in which case the "default" merge option is badly named ..). But that would actually be a change in behaviour.

- The keyword names: `pa.concat_tables` uses `field_merge_options`, `pa.unify_schemas` uses `options`. Should we try to align those names? I mentioned that I would like `promote_options`, which I think is more descriptive, but on the other hand deviates from the C++ termininology ..

The last two bullet points are more about the final Python API, and doesn't impact the large part of the changes in the PR (the actual C++ changes), so maybe I shouldn't block the PR on those, and I am also fine with discussing those in a follow-up.

jorisvandenbossche commented:
And on the `concat_tables` keyword, another idea I mentioned above was to combine the `promote` ("unify_schemas") and `field_merge_options` in a single keyword. This could have options False/"none" (no unification / promotion, only exactly the same schemas are allowed, the default), "default" (unify schemas, with only null to any type promotion) and "permissive" (unify schemas, with permissive type promotion)

Fokko commented:
Thanks @jorisvandenbossche for summarizing the pending issues. I was rebasing to see if the LLVM stuff has been fixed.

> Do we need the pyarrow.FieldMergeOptions class? At the moment it actually doesn't really do anything, since we don't (yet) allow to tweak individual flags, only "permissive"/"default" (and the class itself actually also only exposes "permissive", not "default", as a static method to create that options object). Maybe we can just remove it for now, we can always add it back later if we want to expose individual options.

I don't think it adds much value. Just accepting a string is much clearer to the end-user, and I think the majority of the people don't the fine-grained tuning.

> For concat_tables, I find it confusing that a user still has to specify promote=True when they already specified field_merge_options="permissive" (because the latter only makes sense with promote=True). Essentially the promote=False can be seen as a "no promotion" field merge option?
I think with this new promotion capability, the historically added promote keyword is unfortunately badly chosen. Essentially it means: "unify schemas" (that's also how it is called on the C++ side), and by default it hardly does any promotion (only non-existing fields or null fields to any type). If we keep promote keyword, I would expect that promote=True would do actual, permissive promotion.
To be clear, this is not caused by this PR, but I would still prefer to get a nicer API. One option is to actually rename promote to unify_schemas (which is what it does). We can still keep promote as an alias for now for backwards compatibility.

This is a good point, and I agree that `promote` is poorly chosen. You could have the situation that you don't want to add null fields, but you allow the promotion of the fields. I've added the `none` option as you suggested, and I think this is much cleaner. Also added support for `promote` if people still rely on it. It will emit a warning to let users know that they should switch to `promote_options`.

> The keyword names: pa.concat_tables uses field_merge_options, pa.unify_schemas uses options. Should we try to align those names? I mentioned that I would like promote_options, which I think is more descriptive, but on the other hand deviates from the C++ terminology ..

I like `promote_options` as well. Aligned the names there.

> The last two bullet points are more about the final Python API, and doesn't impact the large part of the changes in the PR (the actual C++ changes), so maybe I shouldn't block the PR on those, and I am also fine with discussing those in a follow-up.

We should not merge anything that we're going to rip out anyway, but I'm happy to help with some of the follow-ups. Getting this one in would be great since it is quite big already.

jorisvandenbossche commented on python/pyarrow/table.pxi:
```diff
@@ -5024,31 +5024,35 @@ def table(data, names=None, schema=None, metadata=None, nthreads=None):
     Concatenate pyarrow.Table objects.
 
-    If promote==False, a zero-copy concatenation will be performed. The schemas
+    If promote_options=="none", a zero-copy concatenation will be performed. The schemas
```
```suggestion
    If promote_options="none", a zero-copy concatenation will be performed. The schemas
```

jorisvandenbossche commented on python/pyarrow/table.pxi:
```diff
@@ -5024,31 +5024,35 @@ def table(data, names=None, schema=None, metadata=None, nthreads=None):
     first table.
 
-    If promote==True, any null type arrays will be casted to the type of other
+    If promote_options=="default", any null type arrays will be casted to the type of other
```
```suggestion
    If promote_options="default", any null type arrays will be casted to the type of other
```

jorisvandenbossche commented on python/pyarrow/table.pxi:
```diff
@@ -5078,11 +5082,24 @@ def concat_tables(tables, c_bool promote=False, MemoryPool memory_pool=None):
 
+    if "promote" in kwargs:
+        warnings.warn(
+            "promote has been superseded by mode='default'.", FutureWarning)
```
```suggestion
            "promote has been superseded by mode='default'.", FutureWarning, stacklevel=2)
```

jorisvandenbossche commented on python/pyarrow/types.pxi:
```diff
@@ -3174,6 +3174,10 @@ def unify_schemas(schemas):
         Schemas to merge into a single one.
+    promote_options : str, default default
+        Accepts strings "default" and "permissive".
+        Default; null and only null can be unified with another type.
+        Permissive; promotes types to the greater common denominator.
```
```suggestion
        Default: null and only null can be unified with another type.
        Permissive: types are promoted to the greater common denominator.
```

jorisvandenbossche commented on python/pyarrow/table.pxi:
```diff
@@ -5078,11 +5082,24 @@ def concat_tables(tables, c_bool promote=False, MemoryPool memory_pool=None):
+        warnings.warn(
+            "promote has been superseded by mode='default'.", FutureWarning)
+        if kwargs['promote'] is True:
+            promote_options = "permissive"
```
```suggestion
            promote_options = "default"
```

Isn't that needed to preserve the current default behaviour?

Fokko replied:
I thought we agreed on promoting when `promote=True`, but this also works for me.

jorisvandenbossche reviewed (commented):
Thanks a lot for the updates! 

Added a few small comments (I can also push if needed to ensure we get this in). 
It might be good to add a small test to ensure `promote` keyword still works.

jorisvandenbossche commented on python/pyarrow/tests/test_table.py:
```diff
@@ -1376,6 +1376,28 @@ def test_concat_tables_with_different_schema_metadata():
+    t2 = pa.Table.from_arrays(
+        [pa.array([1.0, 2.0], type=pa.float32())], ["float_field"])
+
+    result = pa.concat_tables([t1, t2], promote=True)
```
```suggestion
    with pytest.warns(FutureWarning):
        result = pa.concat_tables([t1, t2], promote=True)
```

This asserts the warning is raised and at the same time also ensures we don't unnecessarily see the warning in the pytest logs

jorisvandenbossche commented:
Thanks @Fokko! (and @lidavidm for the initial PR)
